message_sender/tasks.py

```diff
@@ -77,6 +80,71 @@ def run(self, metric_name, metric_value, session=None, **kwargs):
 fire_metric = FireMetric()
 
 
+class ConcurrencyLimiter(object):
+    BUCKET_SIZE = 60
+
+    @classmethod
+    def get_current_message_count(cls, msg_type, delay):
+        # Sum the values in all the buckets to get the total
+        total = 0
+        number_of_buckets = delay // cls.BUCKET_SIZE + 1
+        bucket = int(time.time() // cls.BUCKET_SIZE)
+        for i in range(0, number_of_buckets):
+            value = cache.get(msg_type+"_messages_at_%s" % bucket)
+            bucket = bucket - 1
+            if value:
+                total += int(value)
+        return total
+
+    @classmethod
+    def incr_message_count(cls, msg_type, delay):
+        bucket = int(time.time() // cls.BUCKET_SIZE)
+        key = msg_type + "_messages_at_%s" % bucket
+        value = cache.get(key)
+        if value is None:
+            # Add the bucket size to the expiry time so messages that start at
+            # the end of the bucket still complete
+            cache.set(key, 1, delay + cls.BUCKET_SIZE)
+        else:
+            cache.incr(key)
+
+    @classmethod
+    def decr_message_count(cls, msg_type, msg_time):
+        if msg_type == "voice":
+            delay = getattr(settings, 'VOICE_MESSAGE_DELAY', 0)
+        else:
+            delay = getattr(settings, 'TEXT_MESSAGE_DELAY', 0)
+
+        if not msg_time:
+            return
+
+        # Convert from datetime to seconds since epoch
+        msg_time = (msg_time - datetime(1970, 1, 1)).total_seconds()
+
+        time_since = time.time() - msg_time
+        if time_since > delay:
+            return
+        bucket = int(msg_time // cls.BUCKET_SIZE)
+
+        key = msg_type + "_messages_at_%s" % bucket
+        value = cache.get(key)
+        # Don't allow negative values
+        if value:
+            if value < 0:
+                # Set the expiry time to the delay minus the time passed since
+                # the message was sent.
+                cache.set(key, 0, delay - time_since)
+            else:
+                cache.decr(key)
+
+    @classmethod
+    def manage_limit(cls, task, msg_type, limit, delay):
+        if limit > 0:
+            if cls.get_current_message_count(msg_type, delay) >= limit:
+                task.retry(countdown=delay)
+            cls.incr_message_count(msg_type, delay)
```

Comment:
I think it's something that we want configurable. We don't want to down the line notice that we're waiting for too long, and have to make a code change to change it.

At the moment this setting is only really relevant for voice calls, since text is SMS it doesn't have a concept of sessions. It could be useful with USSD, but I don't think we've ever done push USSD, so maybe if we (for now) just focus the variables for voice calls we'll be okay?